File: app/src/main/java/org/scp/gymlog/ui/common/dialogs/TextSelectDialogFragment.java
package org.scp.gymlog.ui.common.dialogs;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.scp.gymlog.R;

import java.util.List;
import java.util.function.BiConsumer;

public class TextSelectDialogFragment extends DialogFragment {
    public static final int DIALOG_CLOSED = -1;

    private final List<String> texts;
    private final BiConsumer<Integer, String> callback;
    private boolean callbackCalled = false;

    public TextSelectDialogFragment(List<String> texts, BiConsumer<Integer, String> callback) {
        this.texts = texts;
        this.callback = callback;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_menu, null);
        RecyclerView recyclerView = view.findViewById(R.id.parentLayout);

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(new TextSelectRecyclerViewAdapter(texts,
                this::onMenuElementClicked));

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity()).setView(view);

        return builder.create();
    }

    private void onMenuElementClicked(int index, String text) {
        callback.accept(index, text);
        callbackCalled = true;
        dismiss();
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        if (!callbackCalled) {
            callback.accept(DIALOG_CLOSED, null);
        }
        super.onDismiss(dialog);
    }
}

File: app/src/main/java/org/scp/gymlog/ui/common/dialogs/TextSelectDialogFragment.kt
package org.scp.gymlog.ui.common.dialogs

import android.app.AlertDialog
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.scp.gymlog.R
import java.util.function.BiConsumer

class TextSelectDialogFragment(
    private val texts: List<String>,
    private val selectedOption: Int = -1,
    private val callback: BiConsumer<Int, String>,
) : DialogFragment() {

    companion object {
        const val DIALOG_CLOSED = -1
    }

    private var callbackCalled = false

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val inflater = requireActivity().layoutInflater
        val view = inflater.inflate(R.layout.dialog_menu, null)
        val recyclerView: RecyclerView = view.findViewById(R.id.parentLayout)

        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = TextSelectRecyclerViewAdapter(texts, selectedOption) { idx, text ->
            onMenuElementClicked(idx, text)
        }

        val builder = AlertDialog.Builder(activity).setView(view)

        return builder.create()
    }

    private fun onMenuElementClicked(index: Int, text: String) {
        callback.accept(index, text)
        callbackCalled = true
        dismiss()
    }

    override fun onDismiss(dialog: DialogInterface) {
        if (!callbackCalled) {
            callback.accept(DIALOG_CLOSED, "")
        }
        super.onDismiss(dialog)
    }
}
